sql/opt_range.cc

```diff
@@ -7910,6 +7910,20 @@ SEL_TREE *Item_func_ne::get_func_mm_tree(RANGE_OPT_PARAM *param,
 }
 
 
+bool Item_func_between::can_optimize_range(const Item *item,
+                                           bool is_eq_func) const
+{
+  /*
+    BETWEEN cannot allow the range optimization when the argument type is a
+    string and BETWEEN's declared comparison type is not.  This would require
+    a casting operation for the argument.  Rather than cast, disallow the range
+    optimization in this case.
+  */
+  return cmp_type() == item->cmp_type() ||
```

Comment:
Calling Item_func_between::cmp_type() is not relevant here. It is cmp_type() of the Item_func_between itself. It return the BOOLEAN data type and it's cmp_type() is always INT_RESULT, no matter what data types of arguments are.